src/emqtt.erl

```diff
@@ -1215,43 +1305,139 @@ should_ping(ConnMod, Sock, _LastPktId) ->
             Error
     end.
 
-is_inflight_full(#state{max_inflight = infinity}) ->
-    false;
-is_inflight_full(#state{max_inflight = MaxLimit, inflight = Inflight}) ->
-    maps:size(Inflight) >= MaxLimit.
-
-delete_inflight(?PUBACK_PACKET(PacketId, ReasonCode, Properties),
-                State = #state{inflight = Inflight}) ->
-	case maps:find(PacketId, Inflight) of
-        {ok, {publish, #mqtt_msg{packet_id = PacketId}, _Ts}} ->
-            ok = eval_msg_handler(State, puback, #{packet_id   => PacketId,
-                                                   reason_code => ReasonCode,
-                                                   properties  => Properties}),
-            State#state{inflight = maps:remove(PacketId, Inflight)};
+maybe_shoot(State0 = #state{pendings = Pendings, inflight = Inflight}) ->
+    NPendings = drop_expired(Pendings),
+    State = State0#state{pendings = NPendings},
+    case {is_pendings_empty(NPendings), emqtt_inflight:is_full(Inflight)} of
+        {false, false}->
+            shoot(State#state{pendings = NPendings});
+        {true, _} ->
+            {keep_state, State};
+        {_, true} ->
+            {next_state, inflight_full, State}
+    end.
+
+shoot(State = #state{pendings = Pendings}) ->
+    {PubReq, NPendings} = dequeue_publish_req(Pendings),
+    shoot(PubReq, State#state{pendings = NPendings}).
+
+shoot(?PUB_REQ(Msg = #mqtt_msg{qos = ?QOS_0}, _ExpireAt, Callback), State) ->
+    case send(Msg, State) of
+        {ok, NState} ->
+            eval_callback_handler(ok, Callback),
+            maybe_shoot(NState);
+        {error, Reason} ->
+            shutdown(Reason, State)
+    end;
+shoot(?PUB_REQ(Msg = #mqtt_msg{qos = QoS}, ExpireAt, Callback),
+      State = #state{last_packet_id = PacketId, inflight = Inflight})
+  when QoS == ?QOS_1; QoS == ?QOS_2 ->
+    Msg1 = Msg#mqtt_msg{packet_id = PacketId},
+    case send(Msg1, State) of
+        {ok, NState} ->
+            Inflight1 = emqtt_inflight:insert(
```

Comment:
This might now return `error` if inflight is [full](https://github.com/emqx/emqtt/pull/165/files#diff-efa6def01ea123ef7d106ae6e180f6f35f636b34322e2bad9202f01274ad7b4eR64), so we might end up storing that atom instead of the inflight.  There are a couple other similar occurrences below. 

Maybe change the return type to `{ok, inflight} | error` and check the result?

Reply:
This is great advice, I've updated it to the new commit, thanks for pointing it out.